Synth-grid Key Support

cupcake99 opened the pull request:
Hi,

Here's the branched-out synth-grid key settings.

It's the basic set for movement and selection around the grid for the moment. Just to get started with.

I can add more from my experiments (such as track changing via Tab key) after you check this over and see if there's anything you want altered at all.


Byebye.

Commit:
Added Value module and fixed Oscilloscope renderer to show positive and negative correctly

Commit:
Fixed indentation on Value registration"

Commit:
Rename "value" module to "display" module for clarity. Added invert mode switch to oscilloscope, similar to VUmeter module. Add line to gitignore for stupid mac hidden cache files.

Commit:
cleaned up comments

Commit:
Moved value grab into render method instead of cycle. Cleaned up unused variables

Commit:
tidied code and did some streamlining

Commit:
tidied code and did some streamlining

Commit:
Merge branch 'master' of https://github.com/cupcake99/prototracker-modular

i messed up please don't get angry mama

Commit:
managed to work out how to set a class member from a const method, and learned about the ingenious word "mutable"

Commit:
modified display module load initialiser to be clearer in intent, changed code layout to be more standard

Commit:
standardised member variable name and renamed string method to something more descriptive

Commit:
Removed mutable usage and went back to local static variable in the string maker method, general behaviour seems the same

Commit:
Added Xcode project for building on OSX, and updated README.md accordingly.

Commit:
'Xcode made this happen'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Added Lerp module to Xcode project. Reintroduced "debug.h" include in MainEditor.

MainEditor was throwing an error because the debug header was no longer included.

Commit:
Added basic keyboard command set to synth grid.

Key commands are:

- direction keys/pg-up/pg-dwn/home/end = movement around grid
- direction keys plus shift = change value of selected module (if it has parameters)
- return: on empty grid space = opens new module dialogue
        : on container module = opens container
- backspace = go to parent when inside container
- delete = delete selected module

cupcake99 commented:
Also, i've noticed some rounding errors in the const module after save+load. I think it must be to do with the paramscale value when saving/loading a project. It obviously drops precision and then when loading back in sets the const to an odd value. This can lead to the output being off by a small amount. Depending on the value at save it changes the amount the same time each reload. Eg. if the value is 0.65 it loads in as 0.650391, if 0.63 it comes out at 0.630859. This can lead to values being off by a 1000th or even a 100th. It's not catastrophic but definitely can cause some rounding issues. Is there a way to drop the output to just two digits precision, to discard anything after that perhaps? 

I found this out by using a new a module from my wilderness branch called Debug, it's just a debugger that prints module output values to stdout while in runtime.

EDIT: i found a way to fix this. Saving the value from the onDial param as a pseudo-integer (i.e. having no decimal part) and then scaling it down for use works to drop the un-needed digits of precision. 

Like this:
```
void ConstModule::onDial(int delta)
{
	setParam(0, static_cast<float>((delta + static_cast<int>(mParams[0]))));
	snprintf(mLabel, sizeof(mLabel), "%.2f", mParams[0] / 100.0f);
}

void ConstModule::cycle()
{
	// 1.0 * const
	setOutput(0, getParam(0) / 100.0f);

	// 0.001 * const
	setOutput(1, getParam(0) / 100.0f / 1000.0f);
}
```
It seems to be that the problem only occurs when the mParams are written and read from a save file. If they have no floating point beyond .00000 then it removes the inaccuracy.

EDIT EDIT:

BTW this idea breaks backward compatibility with old saves in that the values written for the const modules will be out by a factor of 100 and possibly lost.

kometbomb commented on src/SynthGrid.cpp:
```diff
@@ -569,6 +579,214 @@ bool SynthGrid::onEvent(SDL_Event& event)
 			switch (event.key.keysym.sym)
 			{
+
+				case SDLK_RIGHT:
```
I think we should refactor these, lots of duplicated code. Only the direction changes so this could be something like:

```
case SDLK_LEFT:
case SDLK_RIGHT:
  int direction = event.key.keysym.sym == SDLK_LEFT ? -1 : 1;
...
  mSelectedModule += direction;
...
  ... -> onDial(direction)
...
```

And so on. And similarly to the up/down directions.

cupcake99 replied:
I will get on with this asap.

I should have thought of that really but am not too familiar with using switch statements just yet.

kometbomb replied:
It might be easier to think this as a separate method that is just called in the case.

```
case SDLK_LEFT:
  moveStuff(-1);
  break;
case SDLK_RIGHT:
  moveStuff(1);
  break;
```

kometbomb commented on src/SynthGrid.cpp:
```diff
@@ -569,6 +579,214 @@ bool SynthGrid::onEvent(SDL_Event& event)
+					}
+					else
+					{
+						if (mSelectedModule + 1 <= ModularSynth::maxModules - 1 && (mSelectedModule + 1) % 4 != 0)
```
What's the reasoning behind this?

cupcake99 replied:
It's to stop the selection wrapping around to the next line on the grid. No reason it needs to be there, but i just felt it was nicer that it felt like the grid had 'hard' boundaries. If you removed the part beyond the && it will just wrap each edge over to the next line.

kometbomb replied:
Ah ok, I get it. :+1:

kometbomb replied:
Btw, you will want to check against ```SynthGrid::gridWidth``` instead of 4:

```
if (mSelectedModule < ModularSynth::maxModules - 1 && ((mSelectedModule + 1) % SynthGrid::gridWidth != 0)) 
```

Commit:
Refactored synth grid key switch methods

cupcake99 commented:
Hey,

I refactored the key settings, but yeah, it's still quite long. I agree making a new method might be the best idea.

Will try and work it out. Have a nice weekend.

kometbomb commented:
Good job! I have an idea: I could make a PR on your fork of prototracker-modular with the separate method. Just to try if that makes sense when contributing and sharing ideas.

Have a nice weekend yourself, too. :)

kometbomb committed:
Refactored key shortcuts

kometbomb committed:
Refactored key shortcuts

kometbomb commented:
OK, I was able to commit directly to the branch. Here's my proposition for refactoring, can you please check if it works as you intended.

kometbomb committed:
Added bounds check

cupcake99 commented:
Brilliant, your method is much more concise than what i was able to come up with!

Only thing is that the PgUp/Down no longer jump down to the end row like i had it before. There was a while loop that set the amount to jump if the usual Page amount (+/-16) went out of bounds.

```
if (mSelectedModule + direction < ModularSynth::maxModules - 1)
{
	mSelectedModule += direction;
}
else
{
	int count = 0;

	while (mSelectedModule + count < ModularSynth::maxModules - 4)
	{
		count += 4;
	}

	mSelectedModule = mSelectedModule + count;
}
```

It's nice to have because it gives a quick way to get to the top or bottom row just by a few Pg key moves. If you can come up with a smarter way to get the same effect, fire away! I just know how to do things by brute force computation mostly, and then even most of my efforts are pretty clunky.

SOLVED: _I'm also a bit confused about how to get my local repo in line with what's in the remote. Git isn't letting me pull from the branch for some reason. I've fetched your additions but it tells me the head of my branch is still at the last commit i made (after having reset my HEAD to that one). It's no bother really as i can just delete the local branch once the merge goes into the master. But i do find git confusing sometimessssss._

cupcake99 commented:
Duh i figured it out and needed to do a merge manually... It was a nice straight ffwd down the road and i'm now where you are.

Can't understand why pull didn't just do that for me, as i thought that was its purpose. Anyhoo..

Commit:
Reintroduced variable jump length for cursor on PAGE-UP/DOWN

Now checks for cursor being in a zone where PAGE-UP/DOWN respectively will require a jump less than 16
to stay within bounds, then calculates the number of rows needed for jump.

I spent ages trying to figure this out using modulo and so forth, ended up making a while loop in my
previous method to do this, and all the time didn't realise the simplest way was so obvious. Bah.

cupcake99 commented:
Hey. I figured it out last night. No more while loops or whatnot, just a simple division into a local int variable to count rows to jump.

See what you think, if there's anywhere better to put it perhaps? I couldn't find a decent way to shoehorn the check into the moveCursor method without having to make more complicated if () statements to iteratively pass a value through to the right part etc.


It now behaves exactly the same as before the refactoring. Probably better.. 😉

Commit:
Refactoring: move page-up/down bounds check into moveCursor method

cupcake99 commented:
I was able to move the out-of-bounds check into the moveCursor method now. Managed to make it less complicated than i thought it might be.

kometbomb commented:
I think we can combine the page up and single row up bounds check.
Something like this:
```
if (movingUpOrDown) {
mSelectedModule += delta;
if (mSelectedModule < mSelectedModule % SynthGrid::gridWidth)
  mSelectedModule = mSelectedModule % SynthGrid::gridWidth;

if (mSelectedModule >  SynthGrid::maxModules - mSelectedModule %
SynthGrid::gridWidth)
  mSelectedModule = SynthGrid::maxModules - mSelectedModule %
SynthGrid::gridWidth;
}
```

The idea is that we can define the upper and lower bounds so that when
moving up/down whole rows (or rows * 4) we know
the user doesn't ever want the cursor to move sideways. So we can just
clamp to the max/min adjusted so that it's actually
the same column as the cursor is on currently. Sort of like how the
left/right bounds check works.

2018-06-13 15:39 GMT+03:00 cupcake99 <notifications@github.com>:

> Ah, i was able to move the out-of-bounds check into the moveCursor method
> now. Managed to make it less complicated than i thought it might be.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kometbomb/prototracker-modular/pull/25#issuecomment-396922794>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCK6bB6hSTeI2vDgWWJ6iObfm4LGnWNks5t8QgVgaJpZM4UeXm8>
> .
>

kometbomb committed:
Refactored limits check for horizontal movement

kometbomb commented:
I implemented the change, IMO it's cleaner and works exactly like the original check. I think this PR is about done, do you have anything to add?

cupcake99 commented:
Hey, i'm not sure you understood the purpose of my changes! 

Basically, if pressing Page keys would send the cursor out of bounds, then only let it jump enough rows to end up in the same column that it started in, but on the first/last row. Because otherwise Page keys have no effect if they will go out of bounds and so are only useful when cursor is in the middle of the screen. It makes it seem as if the key isn't working or something. My way means that Page will always do something, even if not jumping the whole 4 * rows.

kometbomb commented:
Oh sorry, that definitely was the intention of my change. Let me fix my fix.

kometbomb committed:
Refactored the bounds check

cupcake99 commented:
Ok, that fix works just as expected!

Looks like the pull is indeed ready to merge now.

kometbomb commented:
Great! Thanks for this, this fills a nice empty spot in usability.

cupcake99 commented:
You're most welcome. I'll work on some more key functionality/usability PRs soon. But for now i want to get on with the music making!

kometbomb commented:
I was planning to add the ability to type in values for ConstModule. I hope you don't have something for that already... do you? ;)

cupcake99 commented:
Ha, i wish i was good enough to even know where to start with something like that!

I've just been working on making some synth partials atm. Done a triangle wave generator and an implementation of the MaxMSP SimpleFM~ abstract so far. Will make a PR if you're interested in that sort of thing at all.

I also made a new module 'Saturate' which is just a waveshaper/distortion based on the really simple tanh() function idea. Mainly to be used as guard for unruly signals going beyond digital maximum on the output.

kometbomb commented:
I am very interested in both, thank you please.
